Model builders API update

razdoburdin opened the pull request:
I want to start discussion about changing the Model builders API. 
The main challenge is support of both scikit-learn style estimators and non-scikit style booster objects.
In the presented PR I added a new class `GBTDAALModel` for python interface in non-scikit style. It allow to hide all worldly syntax from the old API.

In case of using `XGBoost`, for regression tasks user-side code changes from:
`d4p_model = daal4py.get_gbt_model_from_xgboost(booster)`
 `d4p_prediction = daal4py.gbt_regression_prediction().compute(X_test, d4p_model).prediction`
to
`d4p_model = daal4py.mb.convert_model(booster)`
`d4p_prediction =d4p_model.predict()`

In case of classification problem user-side code changes from:
`d4p_model = daal4py.get_gbt_model_from_xgboost(booster)`
`d4p_prediction=daal4py.gbt_classification_prediction(nClasses=n_classes).compute(X_test, d4p_model).prediction`
to
`d4p_model = daal4py.mb.convert_model(booster)`
`d4p_prediction = d4p_model.predict()`

For support of scikit-style estimators, I updated `GBTDAALClassifier `and `GBTDAALRegressor `classes.
One can use them like this (example for  `XGBoost`, regression task):
`from daal4py.sklearn.ensemble import GBTDAALRegressor`
`reg = xgb.XGBRegressor()`
`reg.fit(X, y)`
`d4p_predt = GBTDAALRegressor.convert_model(reg).predict(X)`

napetrov commented:
the only thing that i don't fully like - CamelCase on .GbtModel.  Might be GBTModel would be better?

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +179,33 @@ def gbt_reg_model_builder(n_features, n_iterations):
+    '''
+
+    def __init__(self, model):
+        framework_name = (model.__class__.__module__, model.__class__.__name__)
```
It doesn't look like framework name, better rename to `submodule_name` and `class_name`

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +179,33 @@ def gbt_reg_model_builder(n_features, n_iterations):
+        framework_name = (model.__class__.__module__, model.__class__.__name__)
+        if framework_name == ("lightgbm.basic", "Booster"):
+            self._model, self._n_classes = get_gbt_model_from_lightgbm(model)
+        elif framework_name == ("xgboost.core", "Booster"):
```
Support of xgb.XGBClassifier/Regressor is required. Same for other frameworks if they contain sklearn-like estimators

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +179,33 @@ def gbt_reg_model_builder(n_features, n_iterations):
+        
+        self._is_regression = isinstance(self._model, gbt_regression_model)
+
+    def predict(self, x, resultsToEvaluate = 'computeClassLabels'):
```
In sklearn terms, `predict` outputs class or regression value while `predict_proba` outputs class probabilities. Thus, resultsToEvaluate should be internal for these methods instead of passed argument

razdoburdin replied:
done

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +179,33 @@ def gbt_reg_model_builder(n_features, n_iterations):
+        
+        self._is_regression = isinstance(self._model, gbt_regression_model)
+
+    def predict(self, x, resultsToEvaluate = 'computeClassLabels'):
```
Input validation is required

razdoburdin replied:
done

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +179,33 @@ def gbt_reg_model_builder(n_features, n_iterations):
     '''
     return gbt_regression_model_builder(n_features, n_iterations)
+
+class GBTModel():
```
Should be derived from sklearn BaseEstimator. Since sklearn is runtime dependency, this class should not be compiled during build

```suggestion
class GBTModel(BaseEstimator):
```

razdoburdin replied:
done

inteldimitrius commented:
The rest looks good to me!

napetrov commented on src/gbt_model_builder.pyx:
```diff
@@ -178,34 +178,4 @@ def gbt_reg_model_builder(n_features, n_iterations):
-                return output.probabilities
-            else:
-                raise TypeError(f"Value of resultsToEvaluate = {resultsToEvaluate} is unknown")
+    return gbt_regression_model_builder(n_features, n_iterations)
```
```suggestion
    return gbt_regression_model_builder(n_features, n_iterations)

```

razdoburdin commented:
Dear all,
I have updated the PR. Currently, the scikit-style estimators are supported. Please see the updated description for details.

inteldimitrius commented:
MacOS and Linux CI checks are failing. Is it okay?

Alexsandruss commented:
NaN errors are not the only ones:
```
======================================================================
ERROR: test_gbt_cls_model_create_from_catboost_batch (test_examples.TestExNpyArray)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/tests/test_examples.py", line 96, in testit
    result = self.call(ex)
  File "/home/vsts/work/1/s/tests/test_examples.py", line 269, in call
    return ex.main(readcsv=np_read_csv)
  File "/home/vsts/work/1/s/examples/daal4py/gbt_cls_model_create_from_catboost_batch.py", line 73, in main
    daal_prediction = daal_predict_algo.compute(X_test, daal_model)
TypeError: Argument 'model' has incorrect type (expected daal4py._daal4py.gbt_classification_model, got tuple)
```

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +181,110 @@ def gbt_reg_model_builder(n_features, n_iterations):
     return gbt_regression_model_builder(n_features, n_iterations)
+
+
+class GBTDAALModel:
```
It's not obvious why this class should be in compiled file, not usual python script.

razdoburdin replied:
> It's not obvious why this class should be in compiled file, not usual python script.

What benefits do you expect from moving this class from a compile file?

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +181,110 @@ def gbt_reg_model_builder(n_features, n_iterations):
+        else:
+            return self._predict_classification(X, fptype, resultsToEvaluate)
+
+def gbt_build_model(model):
+    return GBTDAALModel().build_model(model)
```
Why do not use static method in GBTDAALModel instead of separate function?

razdoburdin replied:
> Why do not use static method in GBTDAALModel instead of separate function?

Separate function is less wordily and looks more clear for the end user.

Alexsandruss replied:
It's not obvious what is returned from this function while static method will make association with estimator class
Also, `gbt_build_model` and `GBTDAALModel` use abbreviature while sklearn API/docs prefer [full name](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn-ensemble-gradientboostingclassifier).

Alexsandruss commented on src/gbt_model_builder.pyx:
```diff
@@ -179,3 +181,110 @@ def gbt_reg_model_builder(n_features, n_iterations):
+
+        return predict_result.prediction.ravel()
+
+    def predict(self, X, fptype="float", resultsToEvaluate="computeClassLabels"):
```
Where is `predict_proba` and `predict_log_proba` methods?

razdoburdin replied:
> Where is `predict_proba` and `predict_log_proba` methods?

This class is an analog of booster object from xgboost/lightgbm/catboost. It is not scikit-style estimator. Look [here](https://github.com/intel/scikit-learn-intelex/blob/3f34dcb5b2f97182f279d8bbbb5cb8eb36f2d298/daal4py/sklearn/ensemble/GBTDAAL.py#L191) for scikit-style classifier and regressor.

Alexsandruss replied:
From example:
```python
    # Conversion to daal4py
    daal_model = d4p.gbt_build_model(lgb_model)

    # daal4py prediction
    daal_prediction = daal_model.predict(X_test)
```
Is there a reason why `daal_model` should not be sklearn-like estimator with full set of prediction functions?

razdoburdin replied:
The reason is that the frameworks (xgboost/lightgbm/catboost) have two parallel versions of API: functional and scikit-style. The problem is that one can't made a scikit-estimator from the Booster object from functional API for xgboost and lightgbm. We need to support model builders not only for scikit-style API users, but also for functional API users of the GBT Frameworks. GBTDAALModel is analog of the Booster object from functional API.

Alexsandruss replied:
> GBTDAALModel is analog of the Booster object from functional API

What is the need to mimic Booster object if most of it's methods are not implemented?

razdoburdin replied:
> 

There is no any standard methods sets for Booster. Each framework has it's own set of methods. We need to support Booster-based API, as far as frameworks support it, and some users use it.

razdoburdin commented:
> @razdoburdin - we probably also need examples to be added with new API

I updated the existed examples and added the new ones for scikit-style API.

Alexsandruss commented:
@mergifyio rebase

napetrov commented on examples/sklearnex/gbt_cls_model_create_from_catboost_batch.py:
```diff
@@ -0,0 +1,89 @@
+#===============================================================================
+# Copyright 2021 Intel Corporation
```
Looks that those older names were not removed

razdoburdin replied:
fixed

napetrov commented on examples/sklearnex/model_builders_catboost.py:
```diff
@@ -0,0 +1,89 @@
+#===============================================================================
+# Copyright 2021 Intel Corporation
```
```suggestion
# Copyright 2023 Intel Corporation
```

napetrov commented on examples/sklearnex/model_builders_lightgbm.py:
```diff
@@ -0,0 +1,84 @@
+#===============================================================================
+# Copyright 2020 Intel Corporation
```
```suggestion
# Copyright 2023 Intel Corporation
```

napetrov commented on examples/sklearnex/model_builders_xgboost.py:
```diff
@@ -0,0 +1,85 @@
+#===============================================================================
+# Copyright 2014 Intel Corporation
```
```suggestion
# Copyright 2023 Intel Corporation
```

razdoburdin commented:
/intelci: run

Alexsandruss commented on examples/sklearnex/model_builders_xgboost.py:
```diff
@@ -0,0 +1,85 @@
+    return pd.read_csv(f, usecols=c, delimiter=',', header=None, dtype=t)
+
+
+def main(readcsv=pd_read_csv, method='defaultDense'):
```
Method is not used. Same for second example.
```suggestion
def main(readcsv=pd_read_csv):
```

razdoburdin replied:
> Method is not used. Same for second example.

done

razdoburdin commented on daal4py/sklearn/ensemble/GBTDAAL.py:
```diff
@@ -27,7 +27,7 @@
 from .._utils import getFPType
 
 
-class GBTDAALBase(BaseEstimator):
```
> Implement nan tag dispatching for base class:
> 
> ```python
> def _more_tags(self):
>     return {"allow_nan": self.allow_nan_}
> ```

done

razdoburdin commented on daal4py/mb/ModelBuilders.py:
done

razdoburdin commented:
/intelci: run

napetrov commented:
@razdoburdin  test_model_builders_xgboost/ test_model_builders_lightgbm are failing internally

napetrov commented:
/intelci: run

Alexsandruss commented on src/gbt_convertors.pyx:
```diff
@@ -131,7 +152,8 @@ def get_gbt_model_from_xgboost(booster: Any) -> Any:
     trees_arr = booster.get_dump(dump_format="json")
-    xgb_config = json.loads(booster.save_config())
+    if xgb_config is None:
+        xgb_config =get_xgboost_params(booster)
```
```suggestion
        xgb_config = get_xgboost_params(booster)

```

Alexsandruss commented on daal4py/mb/model_builders.py:
```diff
@@ -0,0 +1,195 @@
+    def __init__(self):
+        pass
+
+    def predict(self, X, fptype="float"):
```
`fptype` should be deduced from input data.

napetrov commented:
/intelci: run

samir-nasibli commented:
@razdoburdin you have changed examples names, removed suffix `_batch`. Examples runner will ignore this such file names https://github.com/intel/scikit-learn-intelex/blob/master/tests/run_examples.py#L219:L222 .
Please fix it.

napetrov commented on tests/run_examples.py:
```diff
@@ -157,9 +157,9 @@ def check_library(rule):
-req_library['gbt_cls_model_create_from_lightgbm_batch.py'] = ['lightgbm']
-req_library['gbt_cls_model_create_from_xgboost_batch.py'] = ['xgboost']
-req_library['gbt_cls_model_create_from_catboost_batch.py'] = ['catboost']
+req_library['model_builders_lightgbm.py'] = ['lightgbm']
+req_library['model_builders_xgboost.py'] = ['xgboost']
+req_library['model_builders_catboost.py'] = ['catboost']
```
No. Batch is strange and miningless suffix. If runner will not pick up them, then launcher should be fixed

samir-nasibli replied:
Let's decide how we can group-by this kind of examples then.
All examples [scikit-learn-intelex/examples/](https://github.com/intel/scikit-learn-intelex/tree/master/examples/) are all grouped logically into 4 groups: `spmd.py`, `streaming.py`, `stream.py`, `batch.py`.
If none of them is suitable, then:
* we will come up with a new suffix for such examples.
* update whole examples runner logic

razdoburdin replied:
I have just made a brief look in the run_examples.py. May we just add some sort of default behavior in case the filename doesn't end on `*spmd.py`, `*streaming.py`, `*stream.py` or `*batch.py`?
As far as I see we could switch to this default behavior for all the `*_batch.py` and `*_stream.py` files. In my mind, it will make naming of examples more clear.

napetrov replied:
@razdoburdin i like that idea more, than using suffixes.  

@samir-nasibli - those are logical only if you look on this from some testing perspective. From end user perspective this looks strange.

I would go with base examples without batch because it make no sense to specify this. On stream/streaming - looks like this should have be single group. And for spmd - let's leave it as is

@Alexsandruss, @KulikovNikita   - your thoughts?

samir-nasibli replied:
> I have just made a brief look in the run_examples.py. May we just add some sort of default behavior in case the filename doesn't end on `*spmd.py`, `*streaming.py`, `*stream.py` or `*batch.py`?

Sure we can. That means that we need update examples runner logic little bit (second option).
